src/uattributes/uattributesvalidator.rs

```diff
@@ -360,193 +250,253 @@ impl UAttributesValidators {
+    /// # Errors
     ///
-    /// # Returns
+    /// Returns an error if any of the following checks fail for the given attributes:
     ///
-    /// Returns a `ValidationResult` that is success or failed with a failure message.
-    fn validate_type(&self, attributes: &UAttributes) -> Result<(), UAttributesError> {
-        match attributes.type_.enum_value() {
-            Err(unknown_code) => Err(UAttributesError::validation_error(format!(
-                "Unknown Message Type [{}]",
-                unknown_code
-            ))),
-            Ok(mt) => match mt {
-                UMessageType::UMESSAGE_TYPE_PUBLISH => Ok(()),
-                _ => Err(UAttributesError::validation_error(format!(
-                    "Wrong Message Type [{}]",
-                    mt.to_type_string()
-                ))),
-            },
-        }
-    }
+    /// * [`UAttributesValidator::validate_type`]
+    /// * [`UAttributesValidator::validate_id`]
+    /// * [`UAttributesValidator::validate_ttl`]
+    /// * [`UAttributesValidator::validate_source`]
+    /// * [`UAttributesValidator::validate_sink`]
+    fn validate(&self, attributes: &UAttributes) -> Result<(), UAttributesError> {
+        let error_message = vec![
+            self.validate_type(attributes),
+            self.validate_id(attributes),
+            self.validate_ttl(attributes),
+            self.validate_source(attributes),
+            self.validate_sink(attributes),
+        ]
+        .into_iter()
+        .filter_map(Result::err)
+        .map(|e| e.to_string())
+        .collect::<Vec<_>>()
+        .join("; ");
 
-    fn validate_source_and_sink(&self, attributes: &UAttributes) -> Result<(), UAttributesError> {
-        if (attributes.source.is_none() && attributes.sink.is_none())
-            || (attributes.source.is_some() && attributes.sink.is_some())
-        {
-            return Err(UAttributesError::validation_error(
-                "Message of type PUBLISH must contain either a source or sink",
-            ));
+        if error_message.is_empty() {
+            Ok(())
+        } else {
+            Err(UAttributesError::validation_error(error_message))
         }
-        self.validate_source(attributes)
-            .and_then(|_| self.validate_sink(attributes))
     }
 }
 
 /// Validate `UAttributes` with type `UMessageType::Request`
 pub struct RequestValidator;
 
-impl UAttributesValidator for RequestValidator {
-    fn type_name(&self) -> &'static str {
-        "UAttributesValidator.Request"
+impl RequestValidator {
+    /// Verifies that a set of attributes contains a valid permission level.
+    ///
+    /// # Errors
+    ///
+    /// If [`UAttributes::permission_level`] contains a value that is less than 1, an error is returned.
+    pub fn validate_permission_level(
+        &self,
+        attributes: &UAttributes,
+    ) -> Result<(), UAttributesError> {
+        if let Some(plevel) = attributes.permission_level {
+            if plevel < 1 {
```

Comment:
Isn't 0 okay?

[link to spec](https://github.com/eclipse-uprotocol/up-spec/blob/5f7ea5974b3d74d74c5f013a98cdde575d8708be/basics/qos.adoc#L43)

![image](https://github.com/eclipse-uprotocol/up-rust/assets/11622119/6fedb0a9-abad-4b53-9f42-9f4b59e3d860)

Reply:
This check is about the *permission level*, not the *priority* ...

Reply:
My bad... :sweat:

Reply:
Okay, but [over here in the spec](https://github.com/eclipse-uprotocol/up-spec/blob/5f7ea5974b3d74d74c5f013a98cdde575d8708be/basics/permissions.adoc#22-service-declaration-of-upls) it says that it can be 0 and that is the default, right?

![image](https://github.com/eclipse-uprotocol/up-rust/assets/11622119/11910ad7-42e1-4f8c-b055-4a2a5e31ba7f)

I brought this up a week or so ago on the PR which was going through that changed this and others to unsigned ints.

Reply:
Hmm, seems like you are right. I will change it accordingly ...